sdk/keyvault/keyvault-keys/src/core/challengeBasedAuthenticationPolicy.ts

```diff
@@ -82,94 +101,82 @@ export class ChallengeBasedAuthenticationPolicy extends BaseRequestPolicy {
    * @param webResource
    */
-  public async sendRequest(
-    webResource: WebResource
-  ): Promise<HttpOperationResponse> {
-    if (!webResource.headers) webResource.headers = new HttpHeaders();
+  public async sendRequest(webResource: WebResource): Promise<HttpOperationResponse> {
+    // Headers must exist in the webResource object.
+    if (!webResource.headers) {
+      webResource.headers = new HttpHeaders();
+    }
 
-    // Ensure that we're about to use a secure connection
+    // Ensure that we're about to use a secure connection.
     if (!webResource.url.startsWith("https:")) {
       throw new Error("The resource address for authorization must use the 'https' protocol.");
     }
 
-    const originalBody = webResource.body;
+    // The next request will happen differently whether we have a challenge or not.
+    let response: HttpOperationResponse;
 
     if (this.challengeCache.challenge == undefined) {
-      // Use a blank to start the challenge
+      // If there's no challenge in cache, a blank body will start the challenge.
+      const originalBody = webResource.body;
       webResource.body = "";
+      response = await this._nextPolicy.sendRequest(webResource);
+      webResource.body = originalBody;
     } else {
-      // or use the cached token if we have one
-      await this.authenticateRequest(webResource);
+      // If we did have a challenge in memory,
+      // we attempt to load the token from the cache into the request before we try to send the request.
+      await this.loadToken(webResource);
+      response = await this._nextPolicy.sendRequest(webResource);
     }
 
-    const response = await this._nextPolicy.sendRequest(webResource);
-
-    if (response.status == 401) {
-      webResource.body = originalBody;
-
-      let www_authenticate = response.headers.get("WWW-Authenticate");
-
-      if (www_authenticate) {
-        // The challenge based authentication will contain both an authorization URI with a token,
-        // and the resource to which that token is valid against (also called the scope).
-        const { authorization, resource } = this.parseWWWAuthenticate(www_authenticate);
-        const challenge = new AuthenticationChallenge(authorization, resource + "/.default")
-
-        if (!challenge.equalTo(this.challengeCache.challenge)) {
-          this.challengeCache.setCachedChallenge(challenge);
-          this.tokenCache.setCachedToken(undefined);
-
-          await this.authenticateRequest(webResource);
-          return this._nextPolicy.sendRequest(webResource);
-        }
-        return response;
-      }
+    // If we don't receive a response with a 401 status code,
+    // then we can assume this response has nothing to do with the challenge authentication process.
+    if (response.status !== 401) {
       return response;
-    } else {
+    }
+
+    // If the response status is 401, we only re-authenticate if the WWW-Authenticate header is present.
```

Comment:
instead of doing the split here, could we extract all the handling of the 401 into its own method and call it after the line `webResource.body = originalBody;` above?

Reply:
Can do!

Reply:
Did do!